src/main/groovy/com/google/protobuf/gradle/plugins/ProtobufPlugin.groovy

```diff
@@ -0,0 +1,86 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle.plugins
+
+import org.gradle.api.Action
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.AppliedPlugin
+
+class ProtobufPlugin implements Plugin<Project> {
+    // any one of these plugins should be sufficient to proceed with applying this plugin
+    private static final List<String> javaPluginOptions = ['java']
+    private static final List<String> androidPluginOptions = ['com.android.application',
+                                                              'com.android.library',
+                                                              'android',
+                                                              'android-library']
+
+    private List<String> pluginsApplied = [];
+    private Project project;
+
+    void apply(final Project project) {
+        this.project = project
+
+        // At least one of the prerequisite plugins must by applied before this plugin can be applied, so
+        // we will use the PluginManager.withPlugin() callback mechanism to delay applying this plugin until
+        // after that has been achieved. If project evaluation completes before one of the prerequisite plugins
+        // has been applied then we will assume that none of prerequisite plugins were specified and we will
+        // throw an Exception to alert the user of this configuration issue.
+        applyWithPrerequisitePlugin(javaPluginOptions, 'com.google.protobuf.java')
```

Comment:
Sorry for not replying sooner.

I took a look at how you link the csharp compile tasks and the proto generation tasks in [ProtobufCSharpPlugin.groovy](https://github.com/ngyukman/protobuf-gradle-plugin/blob/e26aa8d59fa06d7cc068cdeeff8691de767bdc07/src/main/groovy/com/google/protobuf/gradle/plugins/ProtobufCSharpPlugin.groovy#L75). Csharp doesn't seem to have a concept of sourceSet in `build.gradle`, thus you simply link all proto sourceSets to all msbuild compile tasks with a full mesh. So practically only one proto sourceSet would be sufficient. Providing the ability for defining protobuf sourceSets doesn't seem useful.

It appears to me that the msbuild plugin is more like a shim between Gradle and `msbuild`, given that the essential of the build is still from in `.sln` and `.csproj` instead of `build.gradle`. It doesn't match the way how Java/Android is built in Gradle. Because the Protobuf plugin DSL integrates well with the Java/Android DSL, it doesn't fit well with the msbuild plugin.

Besides csharp, there are probably other languages that also lack full support in Gradle. Perhaps we can use a more generic way to simplify the proto generation for all of them, instead of trying to work with each of them individually. You didn't mention to me what the current pain points of using Protobuf plugin with csharp, but I guess it's the fact that Java/Android plugin is required to create the sourceSet. If it's the case, we may just make the Protobuf plugin create sourceSets if Java/Android plugin is not loaded, so that the user can still specify proto sourceSets, and use the `generateProtoTasks` closure to link the generateProto tasks to the language compile tasks, e.g., the msbuild task.